File: iOS/Kalorie/Core/UseCases/CreateFoodItemUseCase.swift
//
//  CreateFoodItemUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation
import FirebaseFirestore

enum CreateFoodItemError: Error {
    case invalidCode
    case invalidName
    case invalidCalories
    case invalidWeight
    case invalidPortion(FoodPortionError)
    case itemAlreadyExists

    init(_ validationError: FoodItemValidationError) {
        self = validationError.mapped(
            invalidCode: .invalidCode,
            invalidName: .invalidName,
            invalidCalories: .invalidCalories,
            invalidWeight: .invalidWeight
        ) { .invalidPortion($0) }
    }
}

protocol CreateFoodItemUseCaseProtocol {
    func callAsFunction(_ item: FoodItemDomain) async throws -> FoodItemDomain
}

struct CreateFoodItemUseCase: CreateFoodItemUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol) {
        self.dataProvider = dataProvider
    }

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain) async throws -> FoodItemDomain {
        if let validationError = FoodItemValidation.validate(item) {
            throw CreateFoodItemError(validationError)
        }
        let existing: FoodItemDTO? = try await dataProvider.loadFromServerAsync(
            id: item.id,
            from: Constants.Firestore.foodItems
        )
        guard existing == nil else { throw CreateFoodItemError.itemAlreadyExists }
        let dto = FoodItemDTO(item: item)
        do {
            try await dataProvider.setAsync(dto, id: item.id, in: Constants.Firestore.foodItems)
        } catch let writeError where writeError.matches(domain: FirestoreErrorDomain, code: FirestoreErrorCode.permissionDenied.rawValue) {
            // Firestore's rule denial doesn't say why; a duplicate is the expected cause, but an
            // expired auth session mid-request would also deny with the same code. Re-read to
            // confirm before relabelling the failure as itemAlreadyExists.
            let confirmedExisting: FoodItemDTO? = try? await dataProvider.loadFromServerAsync(
                id: item.id,
                from: Constants.Firestore.foodItems
            )
            guard confirmedExisting != nil else { throw writeError }
            throw CreateFoodItemError.itemAlreadyExists
        }
        return item
    }
}

#if DEBUG
struct CreateFoodItemUseCaseFake: CreateFoodItemUseCaseProtocol {

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain) async throws -> FoodItemDomain {
        item
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/CreateFoodItemUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemValidation
import antoni.kalorie.core.models.FoodItemValidationError
import antoni.kalorie.core.models.FoodPortionError
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemDTO
import antoni.kalorie.core.networking.loadFromServerAsync
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.isFirestorePermissionDenied

sealed class CreateFoodItemError : Exception() {
    data object InvalidCode : CreateFoodItemError()
    data object InvalidName : CreateFoodItemError()
    data object InvalidCalories : CreateFoodItemError()
    data object InvalidWeight : CreateFoodItemError()
    data class InvalidPortion(val error: FoodPortionError) : CreateFoodItemError()
    data object ItemAlreadyExists : CreateFoodItemError()

    // MARK: - Functions

    companion object {
        fun from(validationError: FoodItemValidationError): CreateFoodItemError = validationError.mapped(
            invalidCode = InvalidCode,
            invalidName = InvalidName,
            invalidCalories = InvalidCalories,
            invalidWeight = InvalidWeight,
        ) { InvalidPortion(it) }
    }
}

interface CreateFoodItemUseCaseProtocol {
    suspend operator fun invoke(item: FoodItemDomain): FoodItemDomain
}

class CreateFoodItemUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
) : CreateFoodItemUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(item: FoodItemDomain): FoodItemDomain {
        FoodItemValidation.validate(item)?.let { throw CreateFoodItemError.from(it) }
        val existing: FoodItemDTO? = dataProvider.loadFromServerAsync(id = item.id, from = Constants.Firestore.FOOD_ITEMS)
        if (existing != null) throw CreateFoodItemError.ItemAlreadyExists
        try {
            dataProvider.setAsync(FoodItemDTO(item), id = item.id, inCollection = Constants.Firestore.FOOD_ITEMS)
        } catch (writeError: Exception) {
            if (!writeError.isFirestorePermissionDenied) throw writeError
            // A rule denial does not say why. A duplicate is the expected cause, but an expired
            // auth session mid-request denies with the same code, so re-read before relabelling.
            val confirmedExisting: FoodItemDTO? = try {
                dataProvider.loadFromServerAsync(id = item.id, from = Constants.Firestore.FOOD_ITEMS)
            } catch (_: Exception) {
                null
            }
            if (confirmedExisting == null) throw writeError
            throw CreateFoodItemError.ItemAlreadyExists
        }
        return item
    }
}
